src/prefect/server/utilities/messaging/__init__.py

```diff
@@ -0,0 +1,157 @@
+import abc
+from contextlib import asynccontextmanager
+import importlib
+from typing import (
+    Any,
+    AsyncContextManager,
+    AsyncGenerator,
+    Awaitable,
+    Callable,
+    Dict,
+    List,
+    Optional,
+    Protocol,
+    Type,
+    TypeVar,
+    Union,
+    runtime_checkable,
+)
+from typing_extensions import Self
+
+from prefect.settings import PREFECT_MESSAGING_CACHE, PREFECT_MESSAGING_BROKER
+from prefect.logging import get_logger
+
+logger = get_logger(__name__)
+
+
+class Message(Protocol):
+    """
+    A protocol representing a message sent to a message broker.
+    """
+
+    data: Union[bytes, str]
+    attributes: Dict[str, Any]
+
+
+M = TypeVar("M", bound=Message)
+
+
+class Cache(abc.ABC):
+    @abc.abstractmethod
+    async def clear_recently_seen_messages(self) -> None:
+        ...
+
+    @abc.abstractmethod
+    async def without_duplicates(self, attribute: str, messages: List[M]) -> List[M]:
+        ...
+
+    @abc.abstractmethod
+    async def forget_duplicates(self, attribute: str, messages: List[M]) -> None:
+        ...
+
+
+class Publisher(abc.ABC):
+    @abc.abstractmethod
+    async def __aenter__(self) -> Self:
+        ...
+
+    @abc.abstractmethod
+    async def __aexit__(self, exc_type, exc_value, traceback):
+        ...
+
+    @abc.abstractmethod
+    async def publish_data(self, data: bytes, attributes: Dict[str, str]):
+        ...
+
+
+MessageHandler = Callable[[Message], Awaitable[None]]
+
+
+class StopConsumer(Exception):
+    """
+    Exception to raise to stop a consumer.
+    """
+
+    def __init__(self, ack: bool = False):
+        self.ack = ack
+
+
+class Consumer(abc.ABC):
+    """
+    Abstract base class for consumers that receive messages from a message broker and
+    call a handler function for each message received.
+    """
+
+    @abc.abstractmethod
+    async def run(self, handler: MessageHandler) -> None:
+        """Runs the consumer (indefinitely)"""
+        ...
+
+
+@runtime_checkable
```

Comment:
@bunchesofdonald I think this turned out pretty nice, what do you think?

Reply:
It's beautiful!